src/transport/websocket.ts

```diff
@@ -0,0 +1,444 @@
+      return new WsSubscription(this, this.pendingSubscriptions.addNew(path, listeners, headers));
+    }
+
+    // Add or select subscription
+    const subID = (!pendingSubID) ?
+      this.subscriptions.addNew(path, listeners, headers) :
+      this.subscriptions.add(pendingSubID, path, listeners, headers);
+
+    this.sendMessage(
+      this.getMessage(
+        SubscribeMessageType,
+        subID,
+        path,
+        headers
+      )
+    );
+
+    return new WsSubscription(this, subID);
+  }
+
+  public unsubscribe(subID: number): void {
+    this.sendMessage(
+      this.getMessage(
+        UnsubscribeMessageType,
+        subID
+      )
+    );
+
+    this.subscriptions.get(subID).listeners.onEnd(null);
+    this.subscriptions.remove(subID);
+  }
+
+  private removeAllSubscribtions () {
+    this.subscriptions.removeAll();
+  }
+
+  private getMessage (
+      messageType: WsMessageType,
+      id: number,
+      path?: string,
+      headers?: ElementsHeaders
+  ): Message {
+    return [
+      messageType,
+      id,
+      path,
+      headers
+    ];
+  }
+
+  private sendMessage (message: Message) {
+    if (this.socket.readyState !== WSReadyState.Open) {
+      return console.warn(`Can't send in "${WSReadyState[this.socket.readyState]}" state`);
+    }
+
+    this.socket.send(JSON.stringify(message));
+  }
+
+  private subscription (subID: number) {
+    return this.subscriptions.get(subID);
+  }
+
+  private recieveMessage (event: any) {
+    this.lastMessageReceivedTimestamp = new Date().getTime();
+
+    // First try to parse event to JSON message.
+    let message: Message;
+    try {
+      message = JSON.parse(event.data);
+    } catch (err) {
+      console.warn(new Error('Message is not valid JSON format'));
+      return;
+    }
+
+    // Validate structure of message
+    const nonValidMessageError = this.validateMessage(message);
+    if (nonValidMessageError) {
+      return console.warn(nonValidMessageError);
+    }
+    
+    const messageType = message.shift();
+
+    switch (messageType) {
+      case PongMessageType:
+        this.onPongMessage(message);
+      return;
+      case PingMessageType:
+        this.onPingMessage(message);
+      return;
+    }
+
+    const subID = message.shift();
+    const subscription = this.subscription(subID);
+
+    if (!subscription) {
+      return console.warn(`Recieved message for non existing subscription id: "${subID}"`);
+    }
+
+    const { listeners } = subscription;
+
+    switch (messageType) {
+      case OpenMessageType:
+        this.onOpenMessage(message, subID, listeners);
+      break;
+      case EventMessageType:
+        this.onEventMessage(message, listeners);
+      break;
+      case EosMessageType:
+        this.onEOSMessage(message, subID, listeners);
+      break;
+      case PongMessageType:
+        this.onPongMessage(message);
+      break;
+      default:
+        listeners.onError(new Error('Recived non existing type of message.'));
```

Comment:
I think in this case the websocket connection (and subscriptions) should be terminated. If we can't parse a message, server and client will potentially get out of sync.